File: Visualization/src/MoveSequence.cpp
#include "MoveSequence.hpp"

MoveSequence::MoveSequence(std::vector<Move*> moves) {
    this->moves = std::deque<Move*>();
    this->undostack = std::stack<Move*>();
    for (Move* move : moves) {
        this->moves.push_back(move);
    }
    this->totalMoves = moves.size();
    this->remainingMoves = moves.size();
    this->currentMove = 0;
    return;
}

Move* MoveSequence::pop() {
    if (this->remainingMoves == 0) { return NULL; }

    Move* move = this->moves.front();
    this->moves.pop_front();
    this->undostack.push(move);

    this->remainingMoves--;
    this->currentMove++;
    
    return move;
}

Move* MoveSequence::undo() {
    if (this->currentMove == 0) { return NULL; }

    Move* move = this->undostack.top();
    this->moves.push_front(move);
    this->undostack.pop();

    this->remainingMoves++;
    this->currentMove--;

    return move;
}

File: WebVis/src/MoveSequence.js
export class MoveSequence {
    constructor(moves = []) {
        this.moves = moves;
        this.undostack = [];
        this.totalMoves = moves.length;
        this.remainingMoves = moves.length;
        this.currentMove = 0;
        this.updateMoveProgressString();
    }

    updateMoveProgressString() {
        this.moveProgressString = `Move #${this.currentMove} / #${this.totalMoves}`;
        document.getElementById("infoOverlay").innerHTML = this.moveProgressString;
    }

    pop() {
        if (this.remainingMoves == 0) {
            return;
        }

        let move = this.moves.shift();

        this.remainingMoves--;
        this.currentMove++;

        this.undostack.push(move);
        
        this.updateMoveProgressString();
        return move;
    }

    undo() {
        if (this.currentMove == 0) {
            return;
        }

        let move = this.undostack.pop();

        this.remainingMoves++;
        this.currentMove--;

        this.moves.unshift(move);

        this.updateMoveProgressString();
        return move.reverse();
    }
}
